From a3a06939cc9f8d86e3390082d4684a595bd0c606 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:20:22 -0400
Subject: [fs] nfsd: don't destroy client if mark_client_expired_locked fails

Message-id: <1410898849-22126-139-git-send-email-bfields@redhat.com>
Patchwork-id: 95753
O-Subject: [Patch RHEL7 138/165] nfsd: don't destroy client if mark_client_expired_locked fails
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Jeff Layton <jlayton@primarydata.com>

If it fails, it means that the client is in use and so destroying it
would be bad. Currently, the client_mutex prevents this from happening
but once we remove it, we won't be able to do this.

Signed-off-by: Jeff Layton <jlayton@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 7abea1e8e81ad7ba100bd97c4881027c73e2db3e
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 56999cb..43e66fc 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -2496,8 +2496,10 @@ nfsd4_create_session(struct svc_rqst *rqstp,
   old = find_confirmed_client_by_name(&unconf->cl_name, nn);
   if (old) {
    status = mark_client_expired_locked(old);
-   if (status)
+   if (status) {
+    old = NULL;
     goto out_free_conn;
+   }
   }
   move_to_confirmed(unconf);
   conf = unconf;
@@ -3041,8 +3043,10 @@ nfsd4_setclientid_confirm(struct svc_rqst *rqstp,
   old = find_confirmed_client_by_name(&unconf->cl_name, nn);
   if (old) {
    status = mark_client_expired_locked(old);
-   if (status)
+   if (status) {
+    old = NULL;
     goto out;
+   }
   }
   move_to_confirmed(unconf);
   conf = unconf;
-- 
1.7.1